### PR TITLE
Fix bindings in comments

### LINKED ISSUE #881: Bindings in comments broken
# Description
Bindings in comments cause the rest of the comment text, including the `-->`, to be rendered to HTML.

See #860 

# Steps to Reproduce

1. Write this code

```ts
import {html, render} from 'lit-html';

render(html`<!-- ${'foo'} bar -->`, document.body);
```

2. See this output...

bar -->

## Live Reproduction Link
https://codepen.io/ruphin/pen/jJVrXd?editors=1010

## Expected Results
Comments remain comments

## Actual Results
See above

## Browsers Affected

- [ ] Chrome
- [ ] Firefox
- [ ] Edge
- [ ] Safari 11
- [ ] Safari 10
- [ ] IE 11

### PR DESCRIPTION
**Pull request by justinfagnani**
Fixes #881 

This is a more incremental fix than #847 or #863. Those may be better in the long term, but we want to measure the perf and size impacts of those first.

### CODE REVIEW

File: src/lib/template-result.ts [hunk 1]
```diff
@@ -43,33 +43,34 @@ export class TemplateResult {
    * Returns a string of HTML used to create a `<template>` element.
    */
   getHTML(): string {
-    const endIndex = this.strings.length - 1;
+    const l = this.strings.length - 1;
     let html = '';
-    for (let i = 0; i < endIndex; i++) {
+    let isTextBinding = true;
```
Comment (jridgewell) [inline src/lib/template-result.ts]: We could change this to look for comment bindings instead?

```js
let isCommentBinding = false;
for (let i = 0; i < l; i++) {
  const s = this.strings[i];
  const open = s.lastIndexOf('<!--');
  isCommentBinding = (open > -1 || isCommentBinding) &&
    s.indexOf('-->', open + 1) === -1;

  if (isCommentBinding) {
    html += s + marker;
    continue;
  }
  const match = lastAttributeNameRegex.exec(s);
  if (match) {
    html += s.substr(0, match.index) + match[1] + match[2] +
      boundAttributeSuffix + match[3] + marker;
  } else {
    html += s + nodeMarker;
  }
}
```

**Comment by justinfagnani** [comment]
@jridgewell it is in the weeds a bit, but I think it's pretty basic to support commenting out HTML inside a template during development. It's quite difficult to do this with JS comments as you have to break up the template literal. Also some IDEs add an HTML comment when running the comment-line command inside an `html`-tagged template literal.

File: src/test/lib/render_test.ts [hunk 2]
```diff
@@ -269,6 +269,12 @@ suite('render()', () => {
       assert.equal(container.querySelector('p')!.textContent, 'baz');
     });
 
+    test('does not inject text from comments with bindings', () => {
+      const t = html`<p>${'foo'}<!-- ${'baz'} ${'baz'} -->${'bar'}</p>`;
```
Comment (jridgewell) [inline src/test/lib/render_test.ts]: We should add tests with a commented-out element with attr bindings, and a child node binding. I don't think the current solution is enough.

Comment (justinfagnani) [reply src/test/lib/render_test.ts]: Good point

File: src/test/lib/render_test.ts [hunk 3]
```diff
@@ -269,12 +269,28 @@ suite('render()', () => {
...
+    test('handles elements in comments with attribute bindings', () => {
+      const t = html`A<!-- <div foo=${'bar'}>B</div> -->C`;
+      render(t, container);
+      // Use innerText instead of textContent because of a crazy bug in
+      // Chrome 41 where textContent would include the textContent of comments!
+      assert.equal(container.innerText, 'AC');
+    });
+
+    test('handles elements in comments with node bindings', () => {
+      const t = html`A<!-- <div>${'B'}</div> -->C`;
```
Comment (jridgewell) [inline src/test/lib/render_test.ts]: Ditto.

File: src/test/lib/render_test.ts [hunk 4]
```diff
@@ -269,12 +269,28 @@ suite('render()', () => {
...
+      const t = html`<p>A<!-- ${'B'} ${'C'} -->D</p>`;
       render(t, container);
       // Use innerText instead of textContent because of a crazy bug in
       // Chrome 41 where textContent would include the textContent of comments!
-      assert.equal(container.querySelector('p')!.innerText, 'foobar');
+      assert.equal(container.querySelector('p')!.innerText, 'AD');
+    });
+
+    test('handles elements in comments with attribute bindings', () => {
+      const t = html`A<!-- <div foo=${'bar'}>B</div> -->C`;
```
Comment (jridgewell) [inline src/test/lib/render_test.ts]: I think you'll need one more binding after the `</div>` to hit the failure case.

**Comment by justinfagnani** [comment]
So far this adds 51 bytes to the bundle size. Not great, but gives us a more realistic baseline to better compare the more invasive template prep rewrite PRs.

File: src/lib/template-result.ts (hunk 1 above)
Comment (justinfagnani) [reply src/lib/template-result.ts]: Thanks! Added this with a couple of changes

**Pull request merged**